staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go

```diff
@@ -36,6 +38,125 @@ const (
 )
 
+// TestKMSPluginLateStart tests the scenario where kms-plugin pod/container starts after kube-apiserver pod/container.
+// Since the Dial to kms-plugin is non-blocking we expect the construction of gRPC service to succeed even when
+// kms-plugin is not yet up - dialing happens in the background.
+func TestKMSPluginLateStart(t *testing.T) {
+	var callTimeout = 3 * time.Second
+
+	service, err := NewGRPCService(endpoint, callTimeout)
+	if err != nil {
+		t.Fatalf("failed to create envelope service, error: %v", err)
+	}
+	defer destroyService(service)
+
+	time.Sleep(callTimeout / 2)
+	server, err := startTestKMSProvider()
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer stopTestKMSProvider(server)
+
+	data := []byte("test data")
+	_, err = service.Encrypt(data)
+	if err != nil {
+		t.Fatalf("failed when execute encrypt, error: %v", err)
+	}
+}
+
+// TestIntermittentConnectionLoss tests the scenario where the connection with kms-plugin is intermittently lost.
+func TestIntermittentConnectionLoss(t *testing.T) {
+	var (
+		wg1      sync.WaitGroup
+		wg2      sync.WaitGroup
+		timeout  = 10 * time.Second
+		blackOut = timeout / 2
+	)
+	// Start KMS Plugin
+	server, err := startTestKMSProvider()
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer stopTestKMSProvider(server)
+
+	//  connect to kms plugin
+	service, err := NewGRPCService(endpoint, timeout)
+	if err != nil {
+		t.Fatalf("failed to create envelope service, error: %v", err)
+	}
+	defer destroyService(service)
+	t.Log("Connected to KMSPlugin")
+
+	// Stop KMS Plugin - simulating connection loss
+	stopTestKMSProvider(server)
+	t.Log("KMS Plugin is stopped")
+
+	wg1.Add(1)
+	wg2.Add(1)
+	go func() {
+		defer wg2.Done()
+		// Call service to encrypt data.
+		data := []byte("test data")
+		t.Log("Sending encrypt request")
+		wg1.Done()
+		_, err := service.Encrypt(data)
+		if err != nil {
+			t.Fatalf("failed when execute encrypt, error: %v", err)
+		}
+	}()
+
+	wg1.Wait()
+	time.Sleep(blackOut)
+	// Start KMS Plugin
+	server, err = startTestKMSProvider()
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer stopTestKMSProvider(server)
+	t.Log("Restarted KMS Plugin")
+
+	wg2.Wait()
+}
+
+func TestUnsupportedVersion(t *testing.T) {
+	ver := "invalid"
+	data := []byte("test data")
+	wantErr := fmt.Errorf(versionErrorf, ver, kmsapiVersion)
+
+	server, err := startTestKMSProvider()
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %ver", err)
+	}
+	defer stopTestKMSProvider(server)
+	setAPIVersion(ver)
+
+	s, err := NewGRPCService(endpoint, 1*time.Second)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer destroyService(s)
+
+	// Encrypt
+	_, err = s.Encrypt(data)
+	if err == nil || err.Error() != wantErr.Error() {
+		t.Errorf("got err: %ver, want: %ver", err, wantErr)
+	}
+
+	destroyService(s)
```

Comment:
why restart the service?

Reply:
Forcing the versionChecked to false - simulating a scenario where decrypt happens to come first.